[RDY] Add interfaces as per UML

Balletie opened the pull request:
~~We should consider adding listeners to the architecture for when a command is done. This way, each command can invoke the listener with either an error result or success result. The problem is that we may want many different kinds of error results, and it varies as to how we want to present these results to the user. Maybe this can be solved with the visitor pattern? See this UML and the embedded note:~~

~~![uml-design](https://cloud.githubusercontent.com/assets/3884925/15098761/fbef6eb4-1544-11e6-8b4e-5f10f89d381c.png)~~

Instead of listeners we can use hooks that are passed to the commands as they are created. This requires the commands to be created from the client end, but this is not a big deal, due to the hooks it stays separated.

![uml-design](https://cloud.githubusercontent.com/assets/3884925/15110997/8986c6b0-15e4-11e6-9050-d515478e796b.png)

Balletie committed:
Add interfaces for part of the architecture as described in UML in docs

Hjdskes commented on shell.client/pom.xml:
```diff
@@ -18,7 +18,25 @@
 
     <dependencies>
         <!-- From here on, all plugins are inherited from the parent project. -->
```
Obviously the deps you added should have been above this line ... :P

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/IDisplay.java:
```diff
@@ -0,0 +1,12 @@
+package org.metaborg.spoofax.shell.client;
+
+/**
+ * An {@link IDisplay} is for displaying the results of a command or expressions that was executed,
```
expression (singular) :P

Hjdskes commented on shell.middle/src/main/java/org/metaborg/spoofax/shell/commands/package-info.java:
```diff
@@ -0,0 +1,5 @@
+ * Contains the logic for executing expressions using the core, or executing commands which are
+ * common accross frontends.
+ */
+package org.metaborg.spoofax.shell.commands;
```
Shouldn't this be `org.metaborg.spoofax.shell.middle.commands`?

Balletie replied:
I thought the client module would also add commands, so that's why I thought that they should have the same package name.

Balletie committed:
checkstyle: Remove access modifiers interface methods

Balletie committed:
Add CommandFactory accepting hooks, and EvaluationCommand

By using hooks, client specific code is separate from the CommandInvoker
logic. Each client can then decide for itself how to display things.

The EvaluationCommand is what gets executed when the input given to the
CommandInvoker does not start with #commandPrefix().

Balletie committed:
Wire together the middle and the client

Make an implementation of the command invoker, which at the moment does
not do much except for executing the evaluationcommand.

Let TerminalEditor be a TerminalUserInterface, by making it implement
IDisplay in addition to IEditor.

Set the onCompleteHook from the Repl class, and invoke the hook with a
dummy evaluation result for testing.

Balletie committed:
Fix CheckStyle warnings with javadoc and formatting

Balletie committed:
FindBugs fixes

Balletie committed:
Checkstyle: trailing whitespace

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/Repl.java:
```diff
@@ -0,0 +1,96 @@
+    public Repl(ICommandInvoker invoker, InputStream in, OutputStream out, OutputStream err)
+        throws IOException {
+        this.invoker = invoker;
+        TerminalUserInterface ui = new TerminalUserInterface(in, out, err);
```
Do we follow Spoofax and use `final` wherever possible? I think we should, but it's up for discussion.

Balletie replied:
yea maybe we should do that.

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/Repl.java:
```diff
@@ -0,0 +1,96 @@
+        throws IOException {
+        this.invoker = invoker;
+        TerminalUserInterface ui = new TerminalUserInterface(in, out, err);
+        ui.setPrompt(coloredFg(Color.RED, "[In ]: "));
```
Small nitpick, but I (and I think others too) associate red with error, so perhaps the default color should be changed to something without an associated meaning. Also, I think we should stick with symbols only, unless we start to get more than just singleline/multiline prompts (my suggestion is to follow Python and use `>>>` for singleline and `...` for multiline)

Balletie replied:
ok

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/Repl.java:
```diff
@@ -0,0 +1,96 @@
+    public void run() throws IOException {
+        System.out.println(Ansi.ansi().a("Welcome to the ").bold().a("Spoofax").reset().a(" REPL"));
+        String input;
+        while (!(input = editor.getInput()).trim().equals("exit")) {
```
We should not forget to remove this hardcoded command -- perhaps add a `TODO` here?

Balletie replied:
I will remove it right away

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+import jline.console.ConsoleReader;
+
+/**
+ * A terminal UI which is both an {@link IEditor} and {@link IDisplay}.
```
Forgot `an` before `{@link IDisplay}`

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * @param in
+     *            The {@link InputStream} from which to read user input
```
Off-topic, but I thought Checkstyle complains about missing periods. :laughing:

Balletie replied:
Apparently only for the main body, not param tags

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+     * @param lastLine
+     *            the line to save.
+     */
+    protected void saveLine(String lastLine) {
```
So your suggestion was to use `protected` where we'd normally use `private`? Perhaps this is something for inclusion in CONTRIBUTING.md.

Balletie replied:
Just use `protected` wherever you may want to override as a subclass. It's only in package scope anyway, so `protected` vs `private` cannot hurt much.

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+     *            The prompt string.
+     */
+    public void setPrompt(String promptString) {
+        prompt = promptString;
```
~~Why not directly use `reader.setPrompt`?~~ Also, can the reader handle a `null` prompt?

Balletie replied:
We can write some tests.

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Override
+    public List<String> history() {
+        // @formatter:off
```
What is this for?

Balletie replied:
so that the formatter does not format the stream functions. Otherwise it does not put each function on a separate line, which I don't want.

Hjdskes replied:
You mean Eclipse's formatter? Can't that be configured?

Balletie replied:
this is just for the Stream API, or other Builder-like patterns. Otherwise I just want the other behaviour. So you can't really configure that.

Hjdskes commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/TerminalUserInterface.java:
```diff
@@ -0,0 +1,141 @@
+    @Override
+    public void displayError(String s) {
+        err.println(s);
+        out.flush();
```
Should be `err.flush()`

Balletie replied:
thanks.

Hjdskes commented on shell.core/src/main/java/org/metaborg/spoofax/shell/core/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * The core of the shell contains the logic of evaluating a given expression in some language.
```
s/of/for/?

Hjdskes commented on shell.middle/pom.xml:
```diff
@@ -18,7 +18,11 @@
+        <dependency>
+            <groupId>org.metaborg.spoofax</groupId>
+            <artifactId>shell.core</artifactId>
+            <version>0.1-SNAPSHOT</version>
```
The `version` tag isn't needed since this is an inherited dependency.

Balletie replied:
the parent pom has a dependency on the shell's core?

Hjdskes replied:
Oh oops! I thought this was a dependency on spoofax itself. Ignore this :P

Hjdskes replied:
In that case, this dep should be placed above the `<!-- From here on, ... -->` line.

Balletie replied:
yea I just did that, sorry

Hjdskes commented on shell.middle/src/main/java/org/metaborg/spoofax/shell/commands/ICommandFactory.java:
```diff
@@ -0,0 +1,22 @@
+     * The parameters that are needed are currently unknown.
+     *
+     * @param onErrorHook Called upon an error by the created {@link IEvaluationCommand}.
+     * @param onCompleteHook Called upon successful by the created {@link IEvaluationCommand}.
```
s/successul/success/g

Hjdskes commented on shell.middle/src/main/java/org/metaborg/spoofax/shell/commands/ICommandInvoker.java:
```diff
@@ -0,0 +1,98 @@
+public interface ICommandInvoker {
+
+    /**
+     * Add a {@link IReplCommand} to be executed when {@code commandName} is given, under the given
```
`an`

Hjdskes commented on shell.middle/src/main/java/org/metaborg/spoofax/shell/commands/ICommandInvoker.java:
```diff
@@ -0,0 +1,98 @@
+     * Ensure that the given parameter is returned without the {@link #commandPrefix()}.
+     *
+     * @param optionallyPrefixedCommandName
+     *            And optionally prefixed command name.
```
An?

Hjdskes commented on shell.middle/src/main/java/org/metaborg/spoofax/shell/commands/SpoofaxCommandFactory.java:
```diff
@@ -0,0 +1,29 @@
+    public void createEvaluationCommand(Consumer<String> onErrorHook,
+                                        Consumer<String> onCompleteHook) {
+        invoker.setEvaluationCommand(s -> {
+            onCompleteHook.accept("Hello.");
```
? :P

Balletie replied:
There's no core yet :P

Unia commented:
Are you going to merge middle and core in this PR?

Balletie committed:
Process comments Jente

Balletie committed:
Further implement and test CommandInvoker.

Also add a CommandNotFoundException to the interface, which is thrown
when a command was not found.

Balletie committed:
Replace hardcoded exit with exit command

Balletie commented:
@Unia, @gfokkema didn't know if we should do that, so you can discuss with him.

gfokkema committed:
Use Guice for factory pattern. (#4)

Balletie commented on shell.client/src/main/java/org/metaborg/spoofax/shell/client/IEditor.java:
```diff
@@ -36,4 +36,20 @@
+     *            The prompt string.
+     */
+	void setContinuationPrompt(String promptString);
+
```
I don't know if we should have these methods in this interface, by the way. But we can keep it like this for now, it's probably fine.

gfokkema replied:
I moved them there since the Repl was attempting to change the prompt, and my changes made that impossible.
My thoughts were that setting a prompt would be universal across editors, but we don't know indeed.

Balletie committed:
Use Guice for dependency injection.

Changes:

Guice is used for dependency injection. Currently this results in a
circular dependency for Repl:

Repl -> Invoker -> commandsMap -> {ExitCommand or OnSuccessHook} -> Repl

Furthermore, Descriptions are now part of Commands, instead of part of
ICommandInvoker.

Balletie commented:
Above PMD error is due to the onError hook that is currently unused, in SpoofaxEvaluationCommand

Balletie committed:
Solve both cyclic dependencies with Guice injection

1st cycle, ExitCommand: Fix by injecting a Provider<Repl> instead of a
  Repl, so that the Repl instance does not have to be available
  immediately upon construction of ExitCommand.

2nd cycle, OnEvalSuccessHook: Fix by injecting an IDisplay instead of a
  Repl. Since TerminalUserInterface is in Singleton scope, this is the
  same instance.

Balletie committed:
Add EvalErrorHook

Balletie committed:
Replace fold with Collectors.joining() for multiline concatenation

Balletie committed:
Update JavaDoc references in EvalCommand

Balletie committed:
Fix JavaDoc link in IEditor

Balletie committed:
Fix comments Jente
